File: kata-setup-typescript/src/GildedRose.ts
import { Item } from "./Item.js"

export class GildedRose {
  items: Array<Item>

  constructor(items = [] as Array<Item>) {
    this.items = items
  }

  updateQuality() {
    this.items.forEach((item) => {
      if (this.isSulfuras(item)) return

      item.decreaseSellIn()

      if (!this.isAgedBrie(item) && !this.isBackstage(item)) {
        item.decreaseQuality()
      } else {
        item.increaseQuality()
        if (this.isBackstage(item)) {
          if (this.isLessThanTenDaysToBackstage(item)) {
            item.increaseQuality()
          }
          if (this.isLessThanFiveDaysToBackstage(item)) {
            item.increaseQuality()
          }
        }
      }
      if (item.isItemOverdated()) {
        if (this.isAgedBrie(item)) {
          item.increaseQuality()
        } else {
          if (this.isBackstage(item)) {
            item.resetQualityToZero()
          } else {
            item.decreaseQuality()
          }
        }
      }
    })

    return this.items
  }

  private isLessThanFiveDaysToBackstage(item: Item) {
    return item.sellIn < 6
  }

  private isLessThanTenDaysToBackstage(item: Item) {
    return item.sellIn < 11
  }

  private isSulfuras(item: Item) {
    return item.name == "Sulfuras, Hand of Ragnaros"
  }

  private isBackstage(item: Item) {
    return item.name == "Backstage passes to a TAFKAL80ETC concert"
  }

  private isAgedBrie(item: Item) {
    return item.name == "Aged Brie"
  }
}

File: kata-setup-java/src/main/java/org/example/GildedRose.java
package org.example;

class GildedRose {

  Item[] items;

  public GildedRose(Item[] items) {
    this.items = items;
  }

  public void updateQuality() {
    for (Item item : items) {
      updateItemQuality(item);
    }
  }

  private static void updateItemQuality(Item item) {
    if (item.isSulfuras()) {
      return;
    }

    if (!item.isAgedBrie()
        && !item.isBackstage()) {
      if (item.canReduceQuality()) {
        item.quality = item.quality - 1;
      }
    } else {
      item.increaseQuality();

      if (item.isBackstage()) {
        if (item.sellIn < 11) {
          item.increaseQuality();
        }

        if (item.sellIn < 6) {
          item.increaseQuality();
        }
      }

    }

    item.sellIn = item.sellIn - 1;

    if (!item.isExpired()) {
      return;
    }

    if (item.isBackstage()) {
      item.quality = Item.MIN_QUALITY;
      return;
    }

    if (item.isAgedBrie()) {
      item.increaseQuality();
    } else {
      if (item.canReduceQuality()) {
        item.quality = item.quality - 1;
      }
    }
  }

}